components/base/wheeled/wheeled_base.go

```diff
@@ -91,17 +100,182 @@ type wheeledBase struct {
+	base := &wheeledBase{
+		widthMm:              cfg.WidthMM,
+		wheelCircumferenceMm: cfg.WheelCircumferenceMM,
+		spinSlipFactor:       cfg.SpinSlipFactor,
+		logger:               logger,
+	}
 
+	if base.spinSlipFactor == 0 {
+		base.spinSlipFactor = 1
+	}
+
+	for _, name := range cfg.Left {
+		m, err := motor.FromDependencies(deps, name)
+		if err != nil {
+			return nil, errors.Wrapf(err, "no left motor named (%s)", name)
+		}
+		positionReporting, err := m.Properties(ctx, nil)
+		if positionReporting["PositonReporting"] && err == nil {
+			base.logger.Debugf("have encoders on motor %#v", m)
+		}
+		base.logger.Debug("no encoders found on left motors")
+		base.left = append(base.left, m)
+	}
+
+	for _, name := range cfg.Right {
+		m, err := motor.FromDependencies(deps, name)
+		if err != nil {
+			return nil, errors.Wrapf(err, "no right motor named (%s)", name)
+		}
+		positionReporting, err := m.Properties(ctx, nil)
+		if positionReporting["PositonReporting"] && err == nil {
+			base.logger.Debugf("have encoders on motor %#v", m)
+		}
+		base.logger.Debug("no encoders found on right motors")
+		base.right = append(base.right, m)
+	}
+
+	for _, msName := range cfg.MovementSensor {
+		ms, err := movementsensor.FromDependencies(deps, msName)
+		if err != nil {
+			return nil, errors.Wrapf(err, "no movement_sensor namesd (%s)", msName)
+		}
+		base.movementSensor = append(base.movementSensor, ms)
+		props, err := ms.Properties(ctx, nil)
+		if props.OrientationSupported && err == nil {
+			base.orienationSupported = props.OrientationSupported
+			base.orientationSensor = ms
+		}
+	}
+
+	base.allMotors = append(base.allMotors, base.left...)
+	base.allMotors = append(base.allMotors, base.right...)
+
+	var err error
+	base.model, err = base.createModelFrame(baseName, cfg.WidthMM)
+	if err != nil {
+		logger.Error("error creating base model, check that a name for the base and width are provided in the configuration")
+	}
+
+	return base, nil
+}
+
+func (base *wheeledBase) createModelFrame(baseName string, widthMM int) (referenceframe.Model, error) {
+	basePose := spatialmath.NewPose(
+		r3.Vector{X: 0, Y: 0, Z: 0},
+		&spatialmath.OrientationVectorDegrees{Theta: 0, OX: 0, OY: 1, OZ: -90})
+	geometry, err := spatialmath.NewBox(
+		basePose,
+		r3.Vector{X: float64(widthMM), Y: float64(widthMM), Z: float64(widthMM)},
+		"base_dimensions")
+	if err != nil {
+		return nil, err
+	}
+	model := referenceframe.NewSimpleModel(baseName)
+	f, err := referenceframe.NewStaticFrameWithGeometry(baseName, basePose, geometry)
+	if err != nil {
+		return nil, err
+	}
+	model.OrdTransforms = append(model.OrdTransforms, f)
+	return model, nil
+}
+
+func (base *wheeledBase) Spin(
+	ctx context.Context, angleDeg, degsPerSec float64, extra map[string]interface{},
+) error {
+	revs, wheelrpm := base.spinMath(angleDeg, degsPerSec)
+
+	switch {
+	case base.orientationSensor != nil && base.orienationSupported:
+		base.logger.Debug("spinning with movement sensor")
```

Comment:
can we put this in a function called spinWithMovementSensor?

Reply:
done